include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,377 @@
+            }
+
+            ~apartment_variable_base() = default;
+
+            // non-copyable, non-assignable
+            apartment_variable_base(apartment_variable_base const&) = delete;
+            void operator=(apartment_variable_base const&) = delete;
+
+            // get current value or throw if no value has been set
+            std::any& get_existing()
+            {
+                if (auto any = get_if())
+                {
+                    return *any;
+                }
+                THROW_HR(E_NOT_SET);
+            }
+
+            // get current value or custom-construct one on demand
+            std::any& get_or_create(const std::function<std::any()>& creator)
+            {
+                const auto apartmentId = test_hook::GetApartmentId();
+
+                auto lock = winrt::slim_lock_guard(s_lock);
+
+                auto storage = s_apartmentStorage.find(apartmentId);
+                if (storage != s_apartmentStorage.end())
+                {
+                    auto& variables = storage->second;
+                    auto variable = variables.find(this);
+                    if (variable != variables.end())
+                    {
+                        return variable->second;
+                    }
+                    else
+                    {
+                        variable = variables.insert({ this, creator() }).first;
+                        return variable->second;
+                    }
+                }
+                else
+                {
+                    struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+                    {
+                        void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                        {
+                            // This code runs at apartment rundown so be careful to avoid deadlocks by
+                            // extracting the variables under the lock then release them outside.
+                            auto variables = [apartmentId]()
+                            {
+                                auto lock = winrt::slim_lock_guard(s_lock);
+                                return s_apartmentStorage.extract(apartmentId);
+                            }();
+                        }
+                    };
+
+                    test_hook::RegisterForApartmentShutdown(winrt::make<ApartmentObserver>().get());
+
+                    auto variables = s_apartmentStorage.insert({ apartmentId,
+                        { { this, creator() } } }).first;
+                    return variables->second.begin()->second; // new map entry is first
+                }
+            }
+
+            // get pointer to current value or nullptr if no value has been set
+            std::any* get_if()
+            {
+                auto lock = winrt::slim_lock_guard(s_lock);
+
+                auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());

```

Comment:
These and other methods could take advantage of the new `get_current_apartment_variables` helper method.